### DISCUSSION TITLE
AvaloniaUI 3D point cloud data visualization

### DISCUSSION THREAD

**Discussion by FSwom** [AvaloniaUI/Avalonia#15776] [Q&A]
Hello,

I am pretty new to 3D visualization, so my question is: 
Can I use Avalonia to read in the 3D point data out of a .pcd file and then visualize them to create something like a PCD viewer? Is this possible with OpenGL or are there other libraries which I can use for this?

**Comment by FSwom** [comment]
After looking a bit for other options I found a library to implement a MonoGame (https://github.com/AvaloniaInside/AvaloniaInside.MonoGame) into the avalonia project. Could this also be an option to create something like a PCD viewer in avalonia?
 
 I appreciate every little help, because I am a bit overwhelmed. 😅

**Reply by jp2masa** [reply]
There are many different approaches. The most common probably are:

- OpenGL using [`OpenGlControlBase`](https://github.com/AvaloniaUI/Avalonia/blob/4fdf4fa36cfc9b1e353a4ca79e5c8e8137bd9ea2/src/Avalonia.OpenGL/Controls/OpenGlControlBase.cs)
- OpenGL, Vulkan, or DirectX using GPU interop API (sample [here](https://github.com/AvaloniaUI/Avalonia/tree/4fdf4fa36cfc9b1e353a4ca79e5c8e8137bd9ea2/samples/GpuInterop), spec [here](#9925))
- Third-party integrations (e.g. `AvaloniaInside.MonoGame`, as you said, not sure if there are others, but a google search should help)

You can also write the integration with a higher-level graphics engine (e.g. game engine) yourself, using either the `OpenGlControlBase` or GPU interop APIs.

**Comment by FSwom** [comment]
Ok, thank you for your help. 

I think I will try to implement it with OpenGL using [OpenGlControlBase](https://github.com/AvaloniaUI/Avalonia/blob/4fdf4fa36cfc9b1e353a4ca79e5c8e8137bd9ea2/src/Avalonia.OpenGL/Controls/OpenGlControlBase.cs). 

Is there anything special I should be aware of with the OpenGL implementation for Avalonia? Or just use the samples like [OpenGLPage](https://github.com/AvaloniaUI/Avalonia/blob/master/samples/ControlCatalog/Pages/OpenGlPage.xaml.cs) and build upon them?

**Reply by jp2masa** [reply]
The only possible problem is the OpenGL implementation on Windows. Avalonia uses ANGLE by default, which lacks some specific features, I think. However, the only alternative is WGL, which is very buggy and not recommended. I recommend testing your app frequently on both Windows and Linux to make sure nothing breaks, but I don't know your target platforms, so this may not apply to your case.

**Reply by FSwom** [reply]
Target platforms are Windows and Linux, so I will try to frequently test it on both platforms to make sure everything works as intended.
